testutil/network.go

```diff
@@ -0,0 +1,337 @@
+	"golang.org/x/sync/errgroup"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	clientkeys "github.com/cosmos/cosmos-sdk/client/keys"
+	"github.com/cosmos/cosmos-sdk/crypto/keyring"
+	"github.com/cosmos/cosmos-sdk/server"
+	"github.com/cosmos/cosmos-sdk/server/api"
+	srvconfig "github.com/cosmos/cosmos-sdk/server/config"
+	"github.com/cosmos/cosmos-sdk/simapp"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	"github.com/cosmos/cosmos-sdk/x/genutil"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+)
+
+var (
+	_, cdc = simapp.MakeCodecs()
+)
+
+// Config defines the necessary configuration used to bootstrap and start an
+// in-process local testing network.
+type Config struct {
+	GenesisState    map[string]json.RawMessage
+	TimeoutCommit   time.Duration
+	ChainID         string
+	NumValidators   int
+	BondDenom       string
+	MinGasPrices    string
+	Passphrase      string
+	AccountTokens   sdk.Int
+	StakingTokens   sdk.Int
+	BondedTokens    sdk.Int
+	PruningStrategy string
+	EnableLogging   bool
+}
+
+// DefaultConfig returns a sane default configuration suitable for nearly all
+// testing requirements.
+func DefaultConfig() Config {
+	return Config{
+		GenesisState:    simapp.ModuleBasics.DefaultGenesis(cdc),
+		TimeoutCommit:   2 * time.Second,
+		ChainID:         "chain-" + tmrand.NewRand().Str(6),
+		NumValidators:   4,
+		BondDenom:       sdk.DefaultBondDenom,
+		MinGasPrices:    fmt.Sprintf("0.000006%s", sdk.DefaultBondDenom),
+		Passphrase:      clientkeys.DefaultKeyPass,
+		AccountTokens:   sdk.TokensFromConsensusPower(1000),
+		StakingTokens:   sdk.TokensFromConsensusPower(500),
+		BondedTokens:    sdk.TokensFromConsensusPower(100),
+		PruningStrategy: storetypes.PruningOptionNothing,
+	}
+}
+
+type (
+	// Network defines a local in-process testing network using SimApp. It can be
+	// configured to start any number of validators, each with its own RPC and API
+	// clients. Typically, this test network would be used in client and integration
+	// testing where user input is expected.
+	Network struct {
+		T          *testing.T
+		BaseDir    string
+		Validators []*Validator
+	}
+
+	// Validator defines an in-process Tendermint validator node. Through this object,
+	// a client can make RPC and API calls and interact with any client command
+	// or handler.
+	Validator struct {
+		AppConfig  *srvconfig.Config
+		ClientCtx  client.Context
+		Ctx        *server.Context
+		Dir        string
+		NodeID     string
+		PubKey     crypto.PubKey
+		Moniker    string
+		APIAddress string
+		RPCAddress string
+		P2PAddress string
+		Address    sdk.AccAddress
+		ValAddress sdk.ValAddress
+		RPCClient  tmclient.Client
+
+		tmNode *node.Node
+		api    *api.Server
+	}
+)
+
+func NewTestNetwork(t *testing.T, cfg Config) *Network {
```

Comment:
we should prob pass a JSONMarshaler instead of using the default from `simapp.MakeCodecs()` 👍

Reply:
It's safer to use the one that is used by the actual underlying application -- SimApp.